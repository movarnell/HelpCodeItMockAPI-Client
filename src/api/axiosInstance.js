// src/api/axiosInstance.js
import axios from 'axios';

const axiosInstance = axios.create({
  baseURL: 'http://localhost:5100', // Adjust if necessary
});

axiosInstance.interceptors.request.use(
    (config) => {
      const token = localStorage.getItem('token');
      console.log('Token from localStorage:', token);
      if (token) {
        config.headers['Authorization'] = `Bearer ${token}`;
      } else {
        console.warn('No token found in localStorage');
      }
      return config;
    },
    (error) => {
      return Promise.reject(error);
    }
  );

export default axiosInstance;