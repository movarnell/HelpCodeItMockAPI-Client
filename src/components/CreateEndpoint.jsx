// src/components/CreateEndpoint.jsx
import React, { useState } from 'react';
import axiosInstance from '../api/axiosInstance';
import { useNavigate, Link } from 'react-router-dom';

function CreateEndpoint() {
  const [endpointName, setEndpointName] = useState('');
  const [httpMethods, setHttpMethods] = useState({
    GET: true,
    POST: true,
    PUT: true,
    DELETE: true,
  });
  const [error, setError] = useState('');
  const navigate = useNavigate();

  const handleSubmit = async (e) => {
    e.preventDefault();
    try {
      const selectedMethods = Object.keys(httpMethods).filter(
        (method) => httpMethods[method]
      );
      console.log('Endpoint Name:', endpointName);
      console.log('Selected Methods:', selectedMethods);

      await axiosInstance.post('/endpoints', {
        endpoint_name: endpointName,
        http_methods: selectedMethods,
      });
      navigate('/endpoints');
    } catch (err) {
      console.error('Error Response:', err.response);
      setError(err.response?.data?.message || 'Failed to create endpoint');
    }
  };

  const handleMethodChange = (e) => {
    const { name, checked } = e.target;
    setHttpMethods((prevMethods) => ({
      ...prevMethods,
      [name]: checked,
    }));
  };

  return (
    <div className="p-6">
      <h1 className="mb-4 text-2xl font-bold">Create New Endpoint</h1>
      {error && <p className="text-red-500">{error}</p>}
      <form onSubmit={handleSubmit} className="space-y-4">
        <div>
          <label className="block text-sm">Endpoint Name</label>
          <input
            type="text"
            className="w-full px-4 py-2 border rounded-md"
            value={endpointName}
            onChange={(e) => setEndpointName(e.target.value)}
            required
          />
          {!endpointName && (
            <p className="text-red-500">Endpoint name is required.</p>
          )}
        </div>
        <div>
          <label className="block text-sm">HTTP Methods</label>
          <div className="flex space-x-4">
            {['GET', 'POST', 'PUT', 'DELETE'].map((method) => (
              <label key={method} className="flex items-center">
                <input
                  type="checkbox"
                  name={method}
                  checked={httpMethods[method]}
                  onChange={handleMethodChange}
                  className="mr-2"
                />
                {method}
              </label>
            ))}
          </div>
          {!Object.values(httpMethods).some(Boolean) && (
            <p className="text-red-500">At least one HTTP method must be selected.</p>
          )}
        </div>
        <button
          type="submit"
          disabled={!endpointName || !Object.values(httpMethods).some(Boolean)}
          className={`px-4 py-2 font-bold text-white bg-blue-500 rounded-md hover:bg-blue-600 ${
            (!endpointName || !Object.values(httpMethods).some(Boolean)) && 'opacity-50 cursor-not-allowed'
          }`}
        >
          Create Endpoint
        </button>
      </form>
      <div className="px-4 py-2 mt-24 mb-4 font-bold text-white bg-green-500 rounded-md hover:bg-green-600 w-fit">
        <Link to={'/dashboard'}> Back to Dashboard</Link>
      </div>
    </div>
  );
}

export default CreateEndpoint;