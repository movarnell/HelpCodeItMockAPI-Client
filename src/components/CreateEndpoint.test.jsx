import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import CreateEndpoint from './CreateEndpoint';
import axiosInstance from '../api/axiosInstance';
import renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';

jest.mock('../api/axiosInstance');
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => jest.fn(),
}));

describe('CreateEndpoint Component', () => {
  const renderComponent = async () => {
    await act(async () => {
      render(
        <BrowserRouter>
          <CreateEndpoint />
        </BrowserRouter>
      );
    });
  };

  test('renders create endpoint form', async () => {
    await renderComponent();
    expect(screen.getByLabelText('Endpoint Name')).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /create endpoint/i })).toBeInTheDocument();
  });

  test('submits form with endpoint data', async () => {
    axiosInstance.post.mockResolvedValue({ data: { endpoint_id: 1 } });
    await renderComponent();

    await act(async () => {
      fireEvent.change(screen.getByLabelText('Endpoint Name'), { target: { value: 'test_endpoint' } });
      fireEvent.click(screen.getByRole('button', { name: /create endpoint/i }));
    });

    await waitFor(() => {
      expect(axiosInstance.post).toHaveBeenCalledWith('/endpoints', {
        endpoint_name: 'test_endpoint',
        http_methods: expect.arrayContaining(['GET', 'POST', 'PUT', 'DELETE'])
      });
    });
  });

  test('matches snapshot', () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <CreateEndpoint />
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});