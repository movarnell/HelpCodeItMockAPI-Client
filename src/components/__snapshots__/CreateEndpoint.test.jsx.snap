// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateEndpoint Component matches snapshot 1`] = `
<div
  className="min-h-screen bg-gray-100"
>
  <nav
    className="bg-white shadow"
  >
    <div
      className="px-4 mx-auto max-w-7xl sm:px-6 lg:px-8"
    >
      <div
        className="flex justify-between h-16"
      >
        <div
          className="flex"
        >
          <div
            className="flex items-center flex-shrink-0"
          >
            <a
              className="text-2xl font-bold text-indigo-600"
              href="/dashboard"
              onClick={[Function]}
            >
              Help Mock It 
            </a>
          </div>
          <div
            className="hidden sm:ml-6 sm:flex sm:space-x-8"
          >
            <a
              className="inline-flex items-center px-1 pt-1 text-sm font-medium border-b-2 border-transparent text-gray-500 hover:border-gray-300 hover:text-gray-700"
              href="/dashboard"
              onClick={[Function]}
            >
              Dashboard
            </a>
            <a
              className="inline-flex items-center px-1 pt-1 text-sm font-medium border-b-2 border-transparent text-gray-500 hover:border-gray-300 hover:text-gray-700"
              href="/endpoints"
              onClick={[Function]}
            >
              Endpoints
            </a>
            <a
              className="inline-flex items-center px-1 pt-1 text-sm font-medium border-b-2 border-transparent text-gray-500 hover:border-gray-300 hover:text-gray-700"
              href="/endpoints/create"
              onClick={[Function]}
            >
              Create Endpoint
            </a>
          </div>
        </div>
        <div
          className="flex items-center"
        >
          <button
            className="inline-flex items-center px-4 py-2 ml-3 text-sm font-medium text-white bg-indigo-600 border border-transparent rounded-md hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500"
            onClick={[Function]}
          >
            Logout
          </button>
        </div>
      </div>
    </div>
  </nav>
  <div
    className="py-6 mx-auto max-w-7xl sm:px-6 lg:px-8"
  >
    <h1
      className="mb-6 text-3xl font-bold text-gray-900"
    >
      Create New Endpoint
    </h1>
    <div
      className="overflow-hidden bg-white shadow sm:rounded-lg"
    >
      <form
        className="p-6 space-y-6"
        onSubmit={[Function]}
      >
        <div>
          <label
            className="block text-sm font-medium text-gray-700"
            htmlFor="endpointName"
          >
            Endpoint Name
          </label>
          <input
            className="block w-full px-3 py-2 mt-1 border border-gray-300 rounded-md shadow-sm focus:outline-none focus:ring-indigo-500 focus:border-indigo-500 sm:text-sm"
            id="endpointName"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
        <div>
          <label
            className="block text-sm font-medium text-gray-700"
          >
            HTTP Methods
          </label>
          <div
            className="mt-2 space-y-2"
          >
            <div
              className="flex items-center"
            >
              <input
                checked={true}
                className="w-4 h-4 text-indigo-600 border-gray-300 rounded focus:ring-indigo-500"
                id="method-GET"
                name="GET"
                onChange={[Function]}
                type="checkbox"
              />
              <label
                className="block ml-2 text-sm text-gray-900"
                htmlFor="method-GET"
              >
                GET
              </label>
            </div>
            <div
              className="flex items-center"
            >
              <input
                checked={true}
                className="w-4 h-4 text-indigo-600 border-gray-300 rounded focus:ring-indigo-500"
                id="method-POST"
                name="POST"
                onChange={[Function]}
                type="checkbox"
              />
              <label
                className="block ml-2 text-sm text-gray-900"
                htmlFor="method-POST"
              >
                POST
              </label>
            </div>
            <div
              className="flex items-center"
            >
              <input
                checked={true}
                className="w-4 h-4 text-indigo-600 border-gray-300 rounded focus:ring-indigo-500"
                id="method-PUT"
                name="PUT"
                onChange={[Function]}
                type="checkbox"
              />
              <label
                className="block ml-2 text-sm text-gray-900"
                htmlFor="method-PUT"
              >
                PUT
              </label>
            </div>
            <div
              className="flex items-center"
            >
              <input
                checked={true}
                className="w-4 h-4 text-indigo-600 border-gray-300 rounded focus:ring-indigo-500"
                id="method-DELETE"
                name="DELETE"
                onChange={[Function]}
                type="checkbox"
              />
              <label
                className="block ml-2 text-sm text-gray-900"
                htmlFor="method-DELETE"
              >
                DELETE
              </label>
            </div>
          </div>
        </div>
        <button
          className="w-full flex justify-center py-2 px-4 border border-transparent rounded-md shadow-sm text-sm font-medium text-white bg-indigo-600 hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500 opacity-50 cursor-not-allowed"
          disabled={true}
          type="submit"
        >
          Create Endpoint
        </button>
      </form>
    </div>
    <div
      className="mt-8"
    >
      <a
        className="inline-flex items-center px-4 py-2 text-sm font-medium text-white bg-green-600 border border-transparent rounded-md hover:bg-green-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-green-500"
        href="/dashboard"
        onClick={[Function]}
      >
        Back to Dashboard
      </a>
    </div>
  </div>
</div>
`;
